src/rate_limiter/src/lib.rs

```diff
@@ -180,24 +190,25 @@ impl TokenBucket {
         }
 
         if tokens > self.budget {
-            // TODO (Issue #259) remove this block when issue is resolved
+            // This operation requests a bandwidth higher than the bucket size
             if tokens > self.size {
                 error!(
                     "Trying to consume more tokens {} than the total capacity {}",
                     tokens, self.size
                 );
-                // best effort rate-limiting, this is a dirty workaround for Issue #259
+                // If the bucket is full, empty it and report an overconsumption of
+                // (tokens / size) times larger than the bucket size
                 if self.budget == self.size {
                     self.budget = 0;
-                    return true;
+                    return BucketReduction::OverConsumption(tokens as f64 / self.size as f64);
                 }
```

Comment:
Instead of waiting for bucket to be full before doing the `OverConsumption` you should do it in place. By waiting for bucket to be full you're also losing precision by not taking the 'waiting' period in the rate-limiting accounting.

```rust
// If the bucket is full, empty it and report an overconsumption of
// (tokens / size) times larger than the bucket size
tokens -= self.budget;
self.budget = 0;
return BucketReduction::OverConsumption(tokens as f64 / self.size as f64);
```

Reply:
Revised the logic in `reduce()` according to this.